File: package/android/src/main/java/com/margelo/nitro/nitromaps/MarkerViewportFilter.kt
package com.margelo.nitro.nitromaps

import com.google.android.gms.maps.model.LatLngBounds
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.sqrt

internal object MarkerViewportFilter {
  /**
   * Precise viewport filter + spatial subsample over pre-narrowed candidates.
   *
   * The caller (spatial index) has already restricted [candidates] to cells
   * near the bounds, so this runs over a small set and is safe to call off the
   * UI thread. Coordinates come from the store's flat arrays.
   */
  fun displaySubset(
    candidates: IntArray,
    latitudes: DoubleArray,
    longitudes: DoubleArray,
    bounds: LatLngBounds,
    latitudeSpan: Double,
  ): IntArray {
    val maxCount = maxMarkersForZoom(latitudeSpan)
    val latSpan = bounds.northeast.latitude - bounds.southwest.latitude
    // Bounds that cross the antimeridian have northeast west of southwest.
    val lngSpan = longitudeSpan(bounds)
    val minLat = bounds.southwest.latitude - latSpan * 0.2
    val maxLat = bounds.northeast.latitude + latSpan * 0.2
    val minLon = wrapLongitude(bounds.southwest.longitude - lngSpan * 0.2)
    val maxLon = wrapLongitude(bounds.northeast.longitude + lngSpan * 0.2)
    val allLongitudes = lngSpan * 1.4 >= 360.0

    val visible = IntList(candidates.size.coerceAtLeast(1))
    for (handle in candidates) {
      val lat = latitudes[handle]
      val lon = longitudes[handle]
      val lonInside = allLongitudes ||
        (if (minLon <= maxLon) lon in minLon..maxLon else lon >= minLon || lon <= maxLon)
      if (lat >= minLat && lat <= maxLat && lonInside) {
        visible.add(handle)
      }
    }

    if (visible.size <= maxCount) {
      return visible.toIntArray()
    }

    return spatialSubsample(visible, latitudes, longitudes, maxCount, bounds)
  }

  private fun spatialSubsample(
    handles: IntList,
    latitudes: DoubleArray,
    longitudes: DoubleArray,
    maxCount: Int,
    bounds: LatLngBounds,
  ): IntArray {
    val columns = ceil(sqrt(maxCount.toDouble())).toInt()
    val rows = ceil(maxCount.toDouble() / columns).toInt()

    val latMin = bounds.southwest.latitude
    val latMax = bounds.northeast.latitude
    val lonMin = bounds.southwest.longitude
    val wraps = bounds.northeast.longitude < lonMin

    val latStep = max(1e-9, (latMax - latMin) / rows)
    val lonStep = max(1e-9, longitudeSpan(bounds) / columns)

    val buckets = LinkedHashMap<Int, IntList>()
    for (index in 0 until handles.size) {
      val handle = handles[index]
      val row = minOf(rows - 1, maxOf(0, ((latitudes[handle] - latMin) / latStep).toInt()))
      // East of the antimeridian the offset from the western edge goes through 180.
      var lonOffset = longitudes[handle] - lonMin
      if (wraps && lonOffset < 0) {
        lonOffset += 360.0
      }
      val column = minOf(columns - 1, maxOf(0, (lonOffset / lonStep).toInt()))
      buckets.getOrPut(row * columns + column) { IntList() }.add(handle)
    }

    val result = IntArray(buckets.size)
    var position = 0
    for (cell in buckets.values) {
      result[position] = cell[cell.size / 2]
      position += 1
    }
    return result
  }

  /** Width of the bounds in degrees, going the short way round the antimeridian. */
  private fun longitudeSpan(bounds: LatLngBounds): Double {
    val raw = bounds.northeast.longitude - bounds.southwest.longitude
    return if (raw < 0) raw + 360.0 else raw
  }

  private fun wrapLongitude(lon: Double): Double {
    var wrapped = lon
    while (wrapped > 180.0) wrapped -= 360.0
    while (wrapped < -180.0) wrapped += 360.0
    return wrapped
  }

  private fun maxMarkersForZoom(latitudeSpan: Double): Int {
    return when {
      latitudeSpan < 0.08 -> 2_000
      latitudeSpan < 0.5 -> 800
      latitudeSpan < 2.0 -> 350
      else -> 200
    }
  }
}

File: package/ios/MarkerViewportFilter.swift
import MapKit

/// Selects a zoom-appropriate marker subset for large datasets.
enum MarkerViewportFilter {
  /// Precise viewport filter + spatial subsample over pre-narrowed candidates.
  ///
  /// The caller (spatial index) has already restricted `candidates` to cells
  /// near the region, so this runs over a small set and is safe to call off the
  /// main thread. Coordinates come from the store's flat arrays.
  static func displaySubset(
    candidates: [Int32],
    latitudes: [Double],
    longitudes: [Double],
    region: MKCoordinateRegion
  ) -> [Int32] {
    let maxCount = maxMarkers(for: region.span.latitudeDelta)
    let bounds = PaddedBounds(region: region, padding: 0.2)
    let visible = candidates.filter { handle in
      bounds.contains(latitude: latitudes[Int(handle)], longitude: longitudes[Int(handle)])
    }

    guard visible.count > maxCount else {
      return visible
    }

    return spatialSubsample(
      visible,
      latitudes: latitudes,
      longitudes: longitudes,
      maxCount: maxCount,
      region: region
    )
  }

  /// Picks at most one marker per geographic cell so subsampling stays visually even.
  private static func spatialSubsample(
    _ handles: [Int32],
    latitudes: [Double],
    longitudes: [Double],
    maxCount: Int,
    region: MKCoordinateRegion
  ) -> [Int32] {
    let columns = Int(ceil(sqrt(Double(maxCount))))
    let rows = Int(ceil(Double(maxCount) / Double(columns)))

    let latMin = region.center.latitude - region.span.latitudeDelta * 0.6
    let latMax = region.center.latitude + region.span.latitudeDelta * 0.6
    let lonMin = region.center.longitude - region.span.longitudeDelta * 0.6
    let lonMax = region.center.longitude + region.span.longitudeDelta * 0.6

    let latStep = max(1e-9, (latMax - latMin) / Double(rows))
    let lonStep = max(1e-9, (lonMax - lonMin) / Double(columns))

    var buckets: [Int: [Int32]] = [:]
    buckets.reserveCapacity(maxCount)

    for handle in handles {
      let index = Int(handle)
      let row = min(rows - 1, max(0, Int((latitudes[index] - latMin) / latStep)))
      let column = min(columns - 1, max(0, Int((longitudes[index] - lonMin) / lonStep)))
      buckets[row * columns + column, default: []].append(handle)
    }

    return buckets.values.map { cell in
      cell[cell.count / 2]
    }
  }

  private static func maxMarkers(for latitudeDelta: Double) -> Int {
    if latitudeDelta < 0.08 {
      return 2_000
    }
    if latitudeDelta < 0.5 {
      return 800
    }
    if latitudeDelta < 2.0 {
      return 350
    }
    return 200
  }

  private struct PaddedBounds {
    let minLat: Double
    let maxLat: Double
    let minLon: Double
    let maxLon: Double

    init(region: MKCoordinateRegion, padding: Double) {
      let latPadding = region.span.latitudeDelta * padding
      let lonPadding = region.span.longitudeDelta * padding
      minLat = region.center.latitude - region.span.latitudeDelta / 2 - latPadding
      maxLat = region.center.latitude + region.span.latitudeDelta / 2 + latPadding
      minLon = region.center.longitude - region.span.longitudeDelta / 2 - lonPadding
      maxLon = region.center.longitude + region.span.longitudeDelta / 2 + lonPadding
    }

    func contains(latitude: Double, longitude: Double) -> Bool {
      let lonInRegion: Bool
      if minLon <= maxLon {
        lonInRegion = longitude >= minLon && longitude <= maxLon
      } else {
        lonInRegion = longitude >= minLon || longitude <= maxLon
      }
      return latitude >= minLat && latitude <= maxLat && lonInRegion
    }
  }
}
